src/validation.rs

```diff
@@ -2186,6 +2194,85 @@ fn verify_distribution_behavior(dist_path: &Path) -> Result<Vec<String>> {
         errors.push("errors running interpreter tests".to_string());
     }
 
+    // Explicitly test ldd directly on the extension modules, which PyInstaller
+    // relies on. This is not strictly needed for a working distribution (e.g.
+    // you can set an rpath on just python+libpython), so we test here for
+    // compatibility with tools that run ldd.
+    // that fails this check (e.g. by setting an rpath on just python+libpython).
+    // https://github.com/pyinstaller/pyinstaller/issues/9204#issuecomment-3171050891
+    // TODO(geofft): musl doesn't do lazy binding for the argument to
+    // ldd, so we will get complaints about missing Py_* symbols. Need
+    // to handle this somehow, skip testing for now.
+    if cfg!(target_os = "linux") && !python_json.target_triple.contains("-musl") {
+        // musl's ldd is packaged in the "musl-tools" Debian package.
+        let ldd = if python_json.target_triple.contains("-musl") && cfg!(not(target_env = "musl")) {
+            "musl-ldd"
+        } else {
+            "ldd"
+        };
+        for (name, variants) in python_json.build_info.extensions.iter() {
+            for ext in variants {
+                let Some(shared_lib) = &ext.shared_lib else {
+                    continue;
+                };
+                let shared_lib_path = temp_dir.path().join("python").join(shared_lib);
+                let output = duct::cmd(ldd, [shared_lib_path])
+                    .unchecked()
+                    .stdout_capture()
+                    .run()
+                    .context(format!("Failed to run `{ldd} {shared_lib}`"))?;
+                let stdout = String::from_utf8_lossy(&output.stdout);
+                // Format of ldd output, for both glibc and musl:
+                // - Everything starts with a tab.
+                // - Most things are "libxyz.so.1 => /usr/lib/libxyz.so.1 (0xabcde000)".
+                // - The ELF interpreter is displayed as just "/lib/ld.so (0xabcde000)".
+                // - glibc, but not musl, shows the vDSO as "linux-vdso.so.1 (0xfffff000)".
```

Comment:
This is not true regarding glibc vs musl for vDSO, just look at the output I showed from Ubuntu and Fedora containers with `ldd`, that's glibc and still includes `linux-vdso.so.1` (these are transitive from glibc).

Realistically, you should only be interested in direct dependencies? As per `patchelf --print-needed` output (_which doesn't add tabbed indents_). That will show what's actually encoded as `DT_NEEDED` entries.

Anything transitive is resolved at runtime, thus `ldd` can be useful for troubleshooting that dependency chain at a runtime environment and filtering for `not found` if anything failed to resolve. Likewise, you could validate the libraries were resolved to the paths you expect.